app/dao/notifications_dao.py

```diff
@@ -707,3 +720,52 @@ def get_service_ids_with_notifications_on_date(notification_type, date):
             union(notification_table_query, ft_status_table_query).subquery()
         ).distinct()
     }
+
+
+def dao_update_delivery_receipts(receipts, delivered):
+    new_receipts = []
+    for r in receipts:
+        if isinstance(r, str):
+            r = json.loads(r)
+        new_receipts.append(r)
+
+    receipts = new_receipts
+
+    id_to_carrier = {
+        r["notification.messageId"]: r["delivery.phoneCarrier"] for r in receipts
+    }
+    id_to_provider_response = {
+        r["notification.messageId"]: r["delivery.providerResponse"] for r in receipts
+    }
+    id_to_timestamp = {r["notification.messageId"]: r["@timestamp"] for r in receipts}
+
+    status_to_update_with = NotificationStatus.DELIVERED
+    if not delivered:
+        status_to_update_with = NotificationStatus.FAILED
+    stmt = (
+        update(Notification)
+        .where(Notification.message_id.in_(id_to_carrier.keys()))
+        .values(
+            carrier=case(
+                *[
+                    (Notification.message_id == key, value)
+                    for key, value in id_to_carrier.items()
+                ]
+            ),
+            status=status_to_update_with,
+            sent_at=case(
+                *[
+                    (Notification.message_id == key, cast(value, TIMESTAMP))
+                    for key, value in id_to_timestamp.items()
+                ]
+            ),
+            provider_response=case(
+                *[
+                    (Notification.message_id == key, value)
+                    for key, value in id_to_provider_response.items()
+                ]
+            ),
+        )
+    )
```

Comment:
I think what you have here is generally what we're looking for, but I want to make sure we're generating the right query/queries under the hood so SQLAlchemy can take full advantage of the underlying DBAPI to make the updates as quick and efficient as possible.

[Take a look here at the note about bulk updates](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#bulk-update), because I think this might need to be restructured so we end up with something like this instead:

```
stmt = (
    update(Notification),
    [
        *notification-1 unpacked as needed,
        *notification-2 unpacked as needed,
        *notification-3 unpacked as needed,
        ...
        *notification-n unpacked as needed,
    ]
)
```

Based on that documentation, it looks like having the primary key as the first thing in the list will help with the filtering that's currently being taken care of with the `where` clause in the code above.

There are other things I've found that folks have tried as well:

* https://stackoverflow.com/questions/54365873/sqlalchemy-update-multiple-rows-in-one-transaction
* https://www.reddit.com/r/flask/comments/lf95rl/how_to_bulk_update_400000_database_entrys_with/

Ultimately, we want to make sure we're letting SQLAlchemy and PostgreSQL do the heavy lifting with this, so let's make sure the query is constructed properly!  This will let us do large batch updates quickly. 🙂

Reply:
I will change to bulk_update_mappings, good catch.

Reply:
As discussed in slack, bulk_update_mappings or some more modern alternative is more complicated because we will have to create some way to map message_ids to notification_ids and insert notification_ids into the delivery receipts.  I will make a ticket and add that to the epic, but for this PR I think we should go with this approach here.

Reply:
Sounds good, thanks again @terrazoon!